src/Titon/Console/Input.hh

```
@@ -0,0 +1,525 @@
+     *
+     * @return $this
+     */
+    public function addCommand(Command $command): this {
+        $command->configure();
+        $this->commands[$command->getName()] = $command;
+
+        return $this;
+    }
+
+    /**
+     * Add a new `Flag` candidate to be parsed from input.
+     *
+     * @param \Titon\Console\Flag $flag The `Flag` to add
+     *
+     * @return $this
+     */
+    public function addFlag(Flag $flag): this {
+        $this->flags->set($flag->getName(), $flag);
+
+        return $this;
+    }
+
+    /**
+     * Add a new `Option` candidate to be parsed from input.
+     *
+     * @param \Titon\Console\Option $option The `Option` to add
+     *
+     * @return $this
+     */
+    public function addOption(Option $option): this {
+        $this->options->set($option->getName(), $option);
+
+        return $this;
+    }
+
+    /**
+     * Parse and retrieve the active command from the raw input.
+     *
+     * @return \Titon\Console\Command|null
+     */
+    public function getActiveCommand(): ?Command {
+        if ($this->parsed === true) {
+            return $this->command;
+        }
+
+        if (!is_null($this->command)) {
+            return $this->command;
+        }
+
+        $input = new Vector($this->rawInput);
+        foreach ($input as $index => $value) {
+            if (!is_null($command = $this->commands->get($value))) {
+                $input->removeKey($index);
+                $this->setInput($input);
+                $this->command = $command;
+
+                return $this->command;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Retrieve an `Argument` by its key or alias. Returns null if none exists.
+     *
+     * @param string $key The key or alias of the `Argument`
+     *
+     * @return \Titon\Console\Argument|null
+     */
+    public function getArgument(string $key): ?Argument {
+        return $this->arguments->get($key);
```

Should we throw an exception here instead of returning null?